src/project.rs

```diff
@@ -0,0 +1,659 @@
+//!
+//! ## The long version
+//!
+//! Let's say you're using a crate which provides a container type like the
+//! following:
+//!
+//! ```rust
+//! /// An unaligned `T`.
+//! ///
+//! /// `Unalign<T>` has the same layout as `T`, except that its alignment
+//! /// is always 1 regardless of `T`'s alignment.
+//! #[repr(C, packed)]
+//! pub struct Unalign<T>{
+//! #    t: T,   
+//! # }
+//! ```
+//!
+//! You're using it with a type from your crate:
+//!
+//! ```rust
+//! #[repr(C)]
+//! struct UdpHeader {
+//!     src_port: u16,
+//!     dst_port: u16,
+//!     length: u16,
+//!     checksum: u16,
+//! }
+//! ```
+//!
+//! Let's say you're reading UDP packets off the network. `UdpHeader` contains
+//! `u16`s, and so (on some platforms) it has an alignment of 2. But you don't
+//! have any control of where your packets are stored in memory, so you can't
+//! construct a `&UdpHeader` from the bytes you've received off the network.
+//! Instead, you construct an `&Unalign<UdpHeader>`.
+//!
+//! That's all well and good, but what if you want to hand out a reference to a
+//! packet's source port? It's a `u16` that might not be validly-aligned, so you
+//! can't just create a `&u16` like you would be able to do if you just had a
+//! `&UdpHeader`. What you'd really like to do is take your
+//! `&Unalign<UdpHeader>` and take a reference to the `src_port`... and get back
+//! an `&Unalign<u16>`. That's what `project!` lets you do:
+//!
+//! ```rust
+//! # use zerocopy::project;
+//! # struct UdpHeader {
+//! #     src_port: u16,
+//! #     dst_port: u16,
+//! #     length: u16,
+//! #     checksum: u16,
+//! # }
+//! # #[repr(C, packed)]
+//! # struct Unalign<T>{
+//! #    t: T,
+//! # }
+//! # unsafe impl<T, F> zerocopy::project::Projectable<F, Unalign<F>> for Unalign<T> {
+//! #     type Inner = T;
+//! # }
+//! # fn read_udp_header() -> &'static Unalign<UdpHeader> { todo!() }
+//! # fn __main() {
+//! let packet = read_udp_header();
+//! // SAFETY: We don't perform projection through any references or other
+//! // fields that implement `Deref` or `DerefMut`.
+//! let src_port = unsafe { project!(&packet.src_port) };
+//! # }
+//! ```
+//!
+//! # How to use this module (for container authors)
+//!
+//! TODO(#196): Fill this section in before removing `#[doc(hidden)]` from this
+//! module.
+
+/// A container which supports field projection of its contained type.
+///
+/// # Example
+///
+/// ```rust
+/// # use zerocopy::project::Projectable;
+/// #[repr(transparent)]
+/// struct Wrapper<T: ?Sized>(T);
+///
+/// unsafe impl<T: ?Sized, F: ?Sized> Projectable<F, Wrapper<F>> for Wrapper<T> {
+///     type Inner = T;
+/// }
+/// ```
+///
+/// # Safety
+///
+/// `Self` must be a `repr(transparent)`, `repr(packed)`, or `repr(C)` struct
+/// with a single non-zero-sized field of type `Self::Inner`. `W` must be a
+/// `repr(transparent)`, `repr(packed)`, or `repr(C)` struct with a single
+/// non-zero field of type `F`.
```

Comment:
Oops, this is just wrong. E.g., it doesn't apply to `MaybeValid<T: AsMaybeUninit>` from #279, whose inner field is not `T`, but `T::MaybeUninit`.

Reply:
Okay, I think this should be [fixed now](https://github.com/google/zerocopy/compare/dcc4affa21f6a30a3fa98e3724ea1c016a6209ce..35847cbd750d79f4401c11bd039a58a7d939d022)?